inspirehep/modules/workflows/tasks/refextract.py

```diff
@@ -191,42 +191,83 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
     )
 
 
-def match_reference(reference):
+def match_references(references):
+    """Match references to their respective records in INSPIRE.
+
+    Args:
+        references (list): the list of references
+
+    Retruns:
+        list: the matched references
+    """
+    matched_references, previous_matched_recid = [], None
+    for ref in references:
+        ref, previous_matched_recid = match_reference(ref, previous_matched_recid)
+        matched_references.append(ref)
+
+    return matched_references
+
+
+def match_reference(reference, previous_matched_recid=None):
     """Match a reference using inspire-matcher.
 
     Args:
         reference (dict): the metadata of a reference.
 
     Returns:
         dict: the matched reference.
-
     """
-    # XXX: avoid this type casting.
-    if get_value(reference, 'reference.publication_info.year'):
-        reference['reference']['publication_info']['year'] = str(reference['reference']['publication_info']['year'])
-
-    config_default = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DEFAULT_CONFIG']
-    config_jcap_and_jhep = current_app.config['WORKFLOWS_REFERENCE_MATCHER_JHEP_AND_JCAP_CONFIG']
-    config_data = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DATA_CONFIG']
+    config_unique_identifiers = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_UNIQUE_IDENTIFIERS_CONFIG']
+    config_default_publication_info = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_DEFAULT_PUBLICATION_INFO_CONFIG']
+    config_jcap_and_jhep_publication_info = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_JHEP_AND_JCAP_PUBLICATION_INFO_CONFIG']
+    config_data = current_app.config[
+        'WORKFLOWS_REFERENCE_MATCHER_DATA_CONFIG']
 
     journal_title = get_value(reference, 'reference.publication_info.journal_title')
-    if journal_title in ['JHEP', 'JCAP']:
-        config = config_jcap_and_jhep
-    else:
-        config = config_default
+    config_publication_info = config_jcap_and_jhep_publication_info if \
+        journal_title in ['JCAP', 'JHEP'] else config_default_publication_info
+
+    configs = [config_unique_identifiers, config_publication_info, config_data]
+
+    matched_reference, matched_recid = \
+        match_reference_with_config(reference, configs, previous_matched_recid)
 
-    result = next(match(reference, config), None)
-    if result:
-        matched_recid = result['_source']['control_number']
-        reference['record'] = get_record_ref(matched_recid, 'literature')
+    return reference, matched_recid
 
-    result_for_data = next(match(reference, config_data), None)
-    if result_for_data:
-        matched_recid = result_for_data['_source']['control_number']
-        reference['record'] = get_record_ref(matched_recid, 'data')
+
+def match_reference_with_config(reference, configs, previous_matched_recid=None):
```

Comment:
this function should take only one config, and instead you should iterate over configs in the caller.